### PR TITLE
Implement escaping to allow clean -p to delete all files when directory contains glob characters

### LINKED ISSUE #10068: cargo clean -p does not clean if path contains glob characters
### Problem

`cargo clean -p` fails to delete a number of files from `target/` if the path to `target` contains glob characters. This is due to the use of `rm_rf_glob` used for cleaning, which does not escape glob characters in the path before adding the pattern:

https://github.com/rust-lang/cargo/blob/e11cd814f8ec5d11a65ae3aea7c3976f250d36a6/src/cargo/ops/cargo_clean.rs#L210-L216

Here for example:

https://github.com/rust-lang/cargo/blob/e11cd814f8ec5d11a65ae3aea7c3976f250d36a6/src/cargo/ops/cargo_clean.rs#L137-L142

### Steps

```bash
#!/bin/bash
rm -rf glob-in-rm
mkdir glob-in-rm
cd glob-in-rm

for d in '[hello]' 'hello'; do
	mkdir "$d"
	pushd "$d"
	cargo new foo
	cd foo
	cargo check
	cargo clean -p foo
	popd
done

diff -qr '[hello]/foo/target' 'hello/foo/target'
```
```
Only in [hello]/foo/target/debug/.fingerprint: foo-8ca57bdc9f429d2a
Only in [hello]/foo/target/debug/deps: foo-8ca57bdc9f429d2a.d
Only in [hello]/foo/target/debug/deps: libfoo-8ca57bdc9f429d2a.rmeta
Only in [hello]/foo/target/debug/incremental: foo-1qp1vfygk93x
```

### Possible Solution(s)

Escape any non-pattern part of the string passed to `glob` with [`glob::Pattern::escape`](https://docs.rs/glob/0.3/glob/struct.Pattern.html#method.escape).

### Notes

_No response_

### Version

```text
cargo 1.56.0 (4ed5d137b 2021-10-04)
release: 1.56.0
commit-hash: 4ed5d137baff5eccf1bae5a7b2ae4b57efad4a7d
commit-date: 2021-10-04
```

### PR DESCRIPTION
**Pull request by russweas**
Closes #10068. 

Runs `glob::Pattern::escape` on path input to `rm_rf_glob()`. This fixes a bug in which `cargo clean -p` fails to delete a number of files from `target/` if the path to target contains glob characters.

### CODE REVIEW

**Comment by jonhoo** [comment]
Ah, you need to make sure to escape only the part of the path that isn't intended to be a pattern. The code that calls `rm_rf_glob` specifically expects one part of the pattern to actually be a glob (something like `{pkg.name()}-*`), and we'll need to keep supporting that (which is probably what the failing test is indicating).

**Comment by russweas** [comment]
> Ah, you need to make sure to escape only the part of the path that isn't intended to be a pattern. The code that calls `rm_rf_glob` specifically expects one part of the pattern to actually be a glob (something like `{pkg.name()}-*`), and we'll need to keep supporting that (which is probably what the failing test is indicating).

Sorry about that! I was under the impression that only I could see draft PR's... I just pushed the correct solution. 

Here's the output of running your script on my machine:
```bash
➜  cargo_test rm -rf glob-in-rm && ./reproduce.sh
~/Projects/cargo_test/glob-in-rm/hello ~/Projects/cargo_test/glob-in-rm
     Created binary (application) `foo` package
    Checking foo v0.1.0 (/Users/russweas/Projects/cargo_test/glob-in-rm/hello/foo)
    Finished dev [unoptimized + debuginfo] target(s) in 0.65s
~/Projects/cargo_test/glob-in-rm
~/Projects/cargo_test/glob-in-rm/[hello] ~/Projects/cargo_test/glob-in-rm
     Created binary (application) `foo` package
    Checking foo v0.1.0 (/Users/russweas/Projects/cargo_test/glob-in-rm/[hello]/foo)
    Finished dev [unoptimized + debuginfo] target(s) in 0.33s
~/Projects/cargo_test/glob-in-rm
Files [hello]/foo/target/.rustc_info.json and hello/foo/target/.rustc_info.json differ
```

**Comment by jonhoo** [comment]
Nice! Draft PRs are visible to everyone, they are just marked as "not ready to merge". The idea of a draft PR is to get feedback on work that's not yet completed, or just to have a place to allow discussion on a fix before it's done. I think it was totally fine to open this as a draft :)

File: src/cargo/core/compiler/layout.rs [hunk 1]
```diff
@@ -164,7 +164,6 @@ impl Layout {
         let lock = dest.open_rw(".cargo-lock", ws.config(), "build directory")?;
         let root = root.into_path_unlocked();
         let dest = dest.into_path_unlocked();
-
```
Comment (jonhoo) [inline src/cargo/core/compiler/layout.rs:167]: This seems like a spurious change that would be best left out to avoid unnecessary conflicts.

Comment (russweas) [reply src/cargo/core/compiler/layout.rs:167]: Have a commit to fix this (and another superfluous formatting change), I think I'll wait to push it though so I don't restart the GH actions...

**Comment by russweas** [comment]
> Nice! Draft PRs are visible to everyone, they are just marked as "not ready to merge". The idea of a draft PR is to get feedback on work that's not yet completed, or just to have a place to allow discussion on a fix before it's done. I think it was totally fine to open this as a draft :)

Now I know 😆. At my internship I used draft PR's like a Todo list of things I was working on, so I would make a tiny change then push it.

**Comment by russweas** [comment]
Note: would also need to incorporate these changes into #10070 
https://github.com/rust-lang/cargo/blob/2f375dab7c563de95ee2d835b0304ae8e307a600/src/cargo/ops/cargo_clean.rs#L176-L184

**Comment by russweas** [comment]
Tests passed, pushed final commit to undo superfluous formatting changes

**Review by jonhoo — CHANGES_REQUESTED** [review]

File: src/cargo/ops/cargo_clean.rs [hunk 2]
```diff
@@ -173,18 +177,21 @@ pub fn clean(ws: &Workspace<'_>, opts: &CleanOptions<'_>) -> CargoResult<()> {
...
+
+                        rm_rf_glob(&dir_glob.join(&hashed_name), config)?;
                         rm_rf(&dir.join(&unhashed_name), config)?;
                         // Remove dep-info file generated by rustc. It is not tracked in
                         // file_types. It does not have a prefix.
-                        let hashed_dep_info = dir.join(format!("{}-*.d", crate_name));
+                        let hashed_dep_info = dir_glob.join(format!("{}-*.d", crate_name));
                         rm_rf_glob(&hashed_dep_info, config)?;
-                        let unhashed_dep_info = dir.join(format!("{}.d", crate_name));
+                        let unhashed_dep_info = dir_glob.join(format!("{}.d", crate_name));
```
[the remark this answers is not in the document]
Comment (russweas) [reply src/cargo/ops/cargo_clean.rs:187]: Good catch

**Comment by alexcrichton** [comment]
Thanks for this! Would you be up for adding some tests for this as well?

**Comment by russweas** [comment]
And just squashed the commits, I noticed it was getting excessive

**Comment by russweas** [comment]
Looks like the test works on `windows-latest, nightly-gnu, i686-pc-windows-gnu` , but not `windows-latest, stable-msvc, i686-pc-windows-msvc`. Specifically, the build artifacts look like they're being generated differently...

**Comment by alexcrichton** [comment]
@bors: r+

Thanks!

**Pull request merged**